In [130]:
import pandas as pd

Panda series: column in a table, 1D 
Panda dataframes: is made up of many panda series - 2D

In [131]:
a = [1,7,2] #list
myvar = pd.Series(a, index=['x', 'y', 'z'])
print(myvar)

x    1
y    7
z    2
dtype: int64


In [132]:
print(myvar[1])
print(myvar["y"])

7
7


In [133]:
#Key/Value Objects as Series
calories = {"day1": 420, "day2": 380, "day3":390}

myvar = pd.Series(calories)
print(myvar)

day1    420
day2    380
day3    390
dtype: int64


In [134]:
#every key will be an index position
print(myvar["day2"])

380


In [135]:
calories = {"day1": 420, "day2": 380, "day3":390}
myvar = pd.Series(calories, index=['day1', 'day2'])
print(myvar)

day1    420
day2    380
dtype: int64


DataFrames: 2D

In [136]:
data = {
    "calories": [420,380,390],
    "duration": [50,40,45]
}

df = pd.DataFrame(data)
print(df)

   calories  duration
0       420        50
1       380        40
2       390        45


In [137]:
data = {
    "calories": [420,380,390],
    "duration": [50,40,45]
}

df = pd.DataFrame(data, index=['day one', 'day two', 'day three'])
print(df)

           calories  duration
day one         420        50
day two         380        40
day three       390        45


boat - Lifeboat (if survived)
body - Body number (if did not survive and body was recovered)

In [138]:
titanic_path = r"C:\Users\12489\Documents\Data\titanic.csv"
titanic = pd.read_csv(titanic_path)

titanic.head()

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

     age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.00      0      0   24160  211.3375       B5        S    2    NaN   
1   0.92      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.00      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

In [139]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   pclass     1309 non-null   int64  
 1   survived   1309 non-null   int64  
 2   name       1309 non-null   object 
 3   sex        1309 non-null   object 
 4   age        1046 non-null   float64
 5   sibsp      1309 non-null   int64  
 6   parch      1309 non-null   int64  
 7   ticket     1309 non-null   object 
 8   fare       1308 non-null   float64
 9   cabin      295 non-null    object 
 10  embarked   1307 non-null   object 
 11  boat       486 non-null    object 
 12  body       121 non-null    float64
 13  home.dest  745 non-null    object 
dtypes: float64(3), int64(4), object(7)
memory usage: 143.3+ KB


In [140]:
#selection of columns and build a separate dataframe
ages = titanic[['age']]
ages.head() 

age
0  29.00
1   0.92
2   2.00
3  30.00
4  25.00

In [141]:
type(ages)

# ages = titanic['age']
# ^^ this is a SERIES ^^

pandas.core.frame.DataFrame

In [142]:
ages.shape

(1309, 1)

In [143]:
age_sex = titanic[['age', 'sex']]
age_sex.head()

age     sex
0  29.00  female
1   0.92    male
2   2.00  female
3  30.00    male
4  25.00  female

In [144]:
type(age_sex)

pandas.core.frame.DataFrame

In [145]:
#conditional filter rows

# mention the condition on a specific column 
# which you want to filter your rows

above_35 = titanic[titanic['age'] > 35]
above_35.head()

pclass  survived                                           name     sex  \
5       1         1                            Anderson, Mr. Harry    male   
6       1         1              Andrews, Miss. Kornelia Theodosia  female   
7       1         0                         Andrews, Mr. Thomas Jr    male   
8       1         1  Appleton, Mrs. Edward Dale (Charlotte Lamson)  female   
9       1         0                        Artagaveytia, Mr. Ramon    male   

    age  sibsp  parch    ticket     fare cabin embarked boat  body  \
5  48.0      0      0     19952  26.5500   E12        S    3   NaN   
6  63.0      1      0     13502  77.9583    D7        S   10   NaN   
7  39.0      0      0    112050   0.0000   A36        S  NaN   NaN   
8  53.0      2      0     11769  51.4792  C101        S    D   NaN   
9  71.0      0      0  PC 17609  49.5042   NaN        C  NaN  22.0   

             home.dest  
5         New York, NY  
6           Hudson, NY  
7          Belfast, NI  
8  Bayside, Queens, NY  
9  Montevideo, Uruguay

In [146]:
above_35.shape

(322, 14)

In [147]:
btwn_35_60 = titanic[(titanic['age'] > 35) & (titanic['age'] < 60)]

btwn_35_60.head()

pclass  survived                                             name     sex  \
5        1         1                              Anderson, Mr. Harry    male   
7        1         0                           Andrews, Mr. Thomas Jr    male   
8        1         1    Appleton, Mrs. Edward Dale (Charlotte Lamson)  female   
10       1         0                           Astor, Col. John Jacob    male   
17       1         1  Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female   

     age  sibsp  parch    ticket      fare    cabin embarked boat   body  \
5   48.0      0      0     19952   26.5500      E12        S    3    NaN   
7   39.0      0      0    112050    0.0000      A36        S  NaN    NaN   
8   53.0      2      0     11769   51.4792     C101        S    D    NaN   
10  47.0      1      0  PC 17757  227.5250  C62 C64        C  NaN  124.0   
17  50.0      0      1  PC 17558  247.5208  B58 B60        C    6    NaN   

              home.dest  
5          New York, NY  
7           Belfast, NI  
8   Bayside, Queens, NY  
10         New York, NY  
17         Montreal, PQ

In [148]:
btwn_35_60.shape

(282, 14)

In [149]:
# isin (categorical) 

# whether each element in the DataFrame is contained 
# in value passenegers from cabin class 2 and 3

class_23 = titanic[titanic['pclass'].isin([2, 3])]

class_23.tail()

pclass  survived                       name     sex   age  sibsp  parch  \
1304       3         0       Zabour, Miss. Hileni  female  14.5      1      0   
1305       3         0      Zabour, Miss. Thamine  female   NaN      1      0   
1306       3         0  Zakarian, Mr. Mapriededer    male  26.5      0      0   
1307       3         0        Zakarian, Mr. Ortin    male  27.0      0      0   
1308       3         0         Zimmerman, Mr. Leo    male  29.0      0      0   

      ticket     fare cabin embarked boat   body home.dest  
1304    2665  14.4542   NaN        C  NaN  328.0       NaN  
1305    2665  14.4542   NaN        C  NaN    NaN       NaN  
1306    2656   7.2250   NaN        C  NaN  304.0       NaN  
1307    2670   7.2250   NaN        C  NaN    NaN       NaN  
1308  315082   7.8750   NaN        S  NaN    NaN       NaN

In [150]:
class_23.shape

(986, 14)

In [151]:
# notna 

# only fetch data for that ISN'T null 
# (notna is a Boolean inverse of isna)

age_no_na = titanic[titanic['age'].notna()]

age_no_na.tail()

pclass  survived                       name     sex   age  sibsp  parch  \
1301       3         0       Youseff, Mr. Gerious    male  45.5      0      0   
1304       3         0       Zabour, Miss. Hileni  female  14.5      1      0   
1306       3         0  Zakarian, Mr. Mapriededer    male  26.5      0      0   
1307       3         0        Zakarian, Mr. Ortin    male  27.0      0      0   
1308       3         0         Zimmerman, Mr. Leo    male  29.0      0      0   

      ticket     fare cabin embarked boat   body home.dest  
1301    2628   7.2250   NaN        C  NaN  312.0       NaN  
1304    2665  14.4542   NaN        C  NaN  328.0       NaN  
1306    2656   7.2250   NaN        C  NaN  304.0       NaN  
1307    2670   7.2250   NaN        C  NaN    NaN       NaN  
1308  315082   7.8750   NaN        S  NaN    NaN       NaN

In [152]:
age_no_na.shape

(1046, 14)

In [153]:
titanic.info()

#  age: 1046 non-null (float64)
# (confirms "age_no_na" above)

# also, as you can see, the original hasn't changed

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   pclass     1309 non-null   int64  
 1   survived   1309 non-null   int64  
 2   name       1309 non-null   object 
 3   sex        1309 non-null   object 
 4   age        1046 non-null   float64
 5   sibsp      1309 non-null   int64  
 6   parch      1309 non-null   int64  
 7   ticket     1309 non-null   object 
 8   fare       1308 non-null   float64
 9   cabin      295 non-null    object 
 10  embarked   1307 non-null   object 
 11  boat       486 non-null    object 
 12  body       121 non-null    float64
 13  home.dest  745 non-null    object 
dtypes: float64(3), int64(4), object(7)
memory usage: 143.3+ KB


## Fetching rows and columns
1. fetch the rows and columns on the basis of index positions: iloc
2. fetch the rows and columns on the basis of labels (row as well as column labels): loc

In [154]:
brics_path = r"C:\Users\12489\Documents\Data\brics.csv"

In [155]:
brics = pd.read_csv(brics_path, index_col=0)

In [156]:
brics

country    capital    area  population
Code                                             
BR          Brazil   Brasilia   8.516      200.40
RU          Russia     Moscow  17.100      143.50
IN           India  New Delhi   3.286     1252.00
CH           China    Beijing   9.597     1357.00
SA    South Africa   Pretoria   1.221       52.98

In [157]:
brics.shape

# "code" is no longer counted as a row

(5, 4)

In [158]:
# type(brics.loc['RU'])

brics.loc[['RU']]

country capital  area  population
Code                                  
RU    Russia  Moscow  17.1       143.5

In [159]:
brics.loc[['RU', 'IN', 'CH']]

country    capital    area  population
Code                                       
RU    Russia     Moscow  17.100       143.5
IN     India  New Delhi   3.286      1252.0
CH     China    Beijing   9.597      1357.0

In [160]:
#select both row and column

brics.loc[['RU', 'IN', 'CH'], ['country', 'capital']]

country    capital
Code                   
RU    Russia     Moscow
IN     India  New Delhi
CH     China    Beijing

In [161]:
brics.loc[:, ['capital', 'area']]

capital    area
Code                   
BR     Brasilia   8.516
RU       Moscow  17.100
IN    New Delhi   3.286
CH      Beijing   9.597
SA     Pretoria   1.221

In [162]:
brics.loc[['BR', 'RU'], :]

country   capital    area  population
Code                                      
BR    Brazil  Brasilia   8.516       200.4
RU    Russia    Moscow  17.100       143.5

In [163]:
# as you can see, it's unchanged

brics

country    capital    area  population
Code                                             
BR          Brazil   Brasilia   8.516      200.40
RU          Russia     Moscow  17.100      143.50
IN           India  New Delhi   3.286     1252.00
CH           China    Beijing   9.597     1357.00
SA    South Africa   Pretoria   1.221       52.98

In [164]:
brics.iloc[[1]]

country capital  area  population
Code                                  
RU    Russia  Moscow  17.1       143.5

In [165]:
brics.iloc[[0,2,4]]

country    capital   area  population
Code                                            
BR          Brazil   Brasilia  8.516      200.40
IN           India  New Delhi  3.286     1252.00
SA    South Africa   Pretoria  1.221       52.98

In [166]:
brics.iloc[[1,3,4], [0,2,3]]

country    area  population
Code                                  
RU          Russia  17.100      143.50
CH           China   9.597     1357.00
SA    South Africa   1.221       52.98

In [167]:
brics.iloc[:, [2,3]]

area  population
Code                    
BR     8.516      200.40
RU    17.100      143.50
IN     3.286     1252.00
CH     9.597     1357.00
SA     1.221       52.98

In [168]:
brics.iloc[[3,4], :]

country   capital   area  population
Code                                           
CH           China   Beijing  9.597     1357.00
SA    South Africa  Pretoria  1.221       52.98

In [169]:
brics.iloc[2:5, :]

country    capital   area  population
Code                                            
IN           India  New Delhi  3.286     1252.00
CH           China    Beijing  9.597     1357.00
SA    South Africa   Pretoria  1.221       52.98

In [170]:
brics

# remember: the original hasn't changed

country    capital    area  population
Code                                             
BR          Brazil   Brasilia   8.516      200.40
RU          Russia     Moscow  17.100      143.50
IN           India  New Delhi   3.286     1252.00
CH           China    Beijing   9.597     1357.00
SA    South Africa   Pretoria   1.221       52.98

In [171]:
# making alterations!!

brics.loc['RU', 'capital'] = "Yaroslavl"
brics

country    capital    area  population
Code                                             
BR          Brazil   Brasilia   8.516      200.40
RU          Russia  Yaroslavl  17.100      143.50
IN           India  New Delhi   3.286     1252.00
CH           China    Beijing   9.597     1357.00
SA    South Africa   Pretoria   1.221       52.98

How to create new columns derived from existing columns

In [172]:
# modify the fare

titanic["modified_fare"] = titanic['fare']/10

In [173]:
titanic.head()

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

     age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.00      0      0   24160  211.3375       B5        S    2    NaN   
1   0.92      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.00      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  modified_fare  
0                     St Louis, MO       21.13375  
1  Montreal, PQ / Chesterville, ON       15.15500  
2  Montreal, PQ / Chesterville, ON       15.15500  
3  Montreal, PQ / Chesterville, ON       15.15500  
4  Montreal, PQ / Chesterville, ON       15.15500

In [174]:
titanic.shape

(1309, 15)

In [175]:
#renaming of columns

titanic_rename = titanic.rename(columns = {'sex':'gender', 'fare':'cost'})
titanic_rename.head()

pclass  survived                                             name  gender  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

     age  sibsp  parch  ticket      cost    cabin embarked boat   body  \
0  29.00      0      0   24160  211.3375       B5        S    2    NaN   
1   0.92      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.00      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  modified_fare  
0                     St Louis, MO       21.13375  
1  Montreal, PQ / Chesterville, ON       15.15500  
2  Montreal, PQ / Chesterville, ON       15.15500  
3  Montreal, PQ / Chesterville, ON       15.15500  
4  Montreal, PQ / Chesterville, ON       15.15500

In [176]:
titanic_rename = titanic_rename.rename(columns = str.upper)
titanic_rename.tail()

PCLASS  SURVIVED                       NAME  GENDER   AGE  SIBSP  PARCH  \
1304       3         0       Zabour, Miss. Hileni  female  14.5      1      0   
1305       3         0      Zabour, Miss. Thamine  female   NaN      1      0   
1306       3         0  Zakarian, Mr. Mapriededer    male  26.5      0      0   
1307       3         0        Zakarian, Mr. Ortin    male  27.0      0      0   
1308       3         0         Zimmerman, Mr. Leo    male  29.0      0      0   

      TICKET     COST CABIN EMBARKED BOAT   BODY HOME.DEST  MODIFIED_FARE  
1304    2665  14.4542   NaN        C  NaN  328.0       NaN        1.44542  
1305    2665  14.4542   NaN        C  NaN    NaN       NaN        1.44542  
1306    2656   7.2250   NaN        C  NaN  304.0       NaN        0.72250  
1307    2670   7.2250   NaN        C  NaN    NaN       NaN        0.72250  
1308  315082   7.8750   NaN        S  NaN    NaN       NaN        0.78750

In [177]:
# summary statistics
# (remember: only works for numerical columns)

titanic.describe()

pclass     survived          age        sibsp        parch  \
count  1309.000000  1309.000000  1046.000000  1309.000000  1309.000000   
mean      2.294882     0.381971    29.881138     0.498854     0.385027   
std       0.837836     0.486055    14.413493     1.041658     0.865560   
min       1.000000     0.000000     0.170000     0.000000     0.000000   
25%       2.000000     0.000000    21.000000     0.000000     0.000000   
50%       3.000000     0.000000    28.000000     0.000000     0.000000   
75%       3.000000     1.000000    39.000000     1.000000     0.000000   
max       3.000000     1.000000    80.000000     8.000000     9.000000   

              fare        body  modified_fare  
count  1308.000000  121.000000    1308.000000  
mean     33.295479  160.809917       3.329548  
std      51.758668   97.696922       5.175867  
min       0.000000    1.000000       0.000000  
25%       7.895800   72.000000       0.789580  
50%      14.454200  155.000000       1.445420  
75%      31.275000  256.000000       3.127500  
max     512.329200  328.000000      51.232920

In [178]:
titanic[["age", "fare"]].mean()

age     29.881138
fare    33.295479
dtype: float64

In [179]:
titanic[["age", "fare"]].median()

age     28.0000
fare    14.4542
dtype: float64

In [180]:
# isolate columns & statistics

titanic.agg({'age': ['min', 'max', 'median'], 'fare': ['min', 'max', 'mean', 'median']})

age        fare
min      0.17    0.000000
max     80.00  512.329200
median  28.00   14.454200
mean      NaN   33.295479

In [181]:
# groupby function

# (aggregate statitics on the basis of category)

titanic[['sex', 'age']].groupby("sex").mean()

age
sex              
female  28.687088
male    30.585228

In [182]:
titanic.groupby('sex') [['age', 'fare']].mean()

age       fare
sex                         
female  28.687088  46.198097
male    30.585228  26.154601

In [183]:
#grouping multiple columns

titanic.groupby(['sex', 'pclass']) [["fare", "age"]].median()

fare   age
sex    pclass                
female 1       80.92915  36.0
       2       23.00000  28.0
       3       10.48960  22.0
male   1       49.50420  42.0
       2       13.00000  29.5
       3        7.89580  25.0

In [184]:
# notice the difference

titanic.groupby(['pclass', 'sex']) [["age", "fare"]].mean()

age        fare
pclass sex                          
1      female  37.037594  109.412385
       male    41.029272   69.888385
2      female  27.499223   23.234827
       male    30.815380   19.904946
3      female  22.185329   15.324250
       male    25.962264   12.415462

In [185]:
# isolating the unique values

titanic["pclass"].unique()

array([1, 2, 3], dtype=int64)

In [186]:
# counting the unique values

titanic["pclass"].value_counts()

3    709
1    323
2    277
Name: pclass, dtype: int64

In [187]:
#sort the data

titanic.sort_values(by='age', ascending = False).head()

pclass  survived                                               name  \
14         1         1               Barkworth, Mr. Algernon Henry Wilson   
61         1         1  Cavendish, Mrs. Tyrell William (Julia Florence...   
1235       3         0                                Svensson, Mr. Johan   
135        1         0                          Goldschmidt, Mr. George B   
9          1         0                            Artagaveytia, Mr. Ramon   

         sex   age  sibsp  parch    ticket     fare cabin embarked boat  body  \
14      male  80.0      0      0     27042  30.0000   A23        S    B   NaN   
61    female  76.0      1      0     19877  78.8500   C46        S    6   NaN   
1235    male  74.0      0      0    347060   7.7750   NaN        S  NaN   NaN   
135     male  71.0      0      0  PC 17754  34.6542    A5        C  NaN   NaN   
9       male  71.0      0      0  PC 17609  49.5042   NaN        C  NaN  22.0   

                    home.dest  modified_fare  
14              Hessle, Yorks        3.00000  
61    Little Onn Hall, Staffs        7.88500  
1235                      NaN        0.77750  
135              New York, NY        3.46542  
9         Montevideo, Uruguay        4.95042

In [188]:
titanic.sort_values(by=['pclass', 'age'], ascending=[False, True]).head()

pclass  survived                                     name     sex   age  \
763        3         1  Dean, Miss. Elizabeth Gladys "Millvina"  female  0.17   
747        3         0  Danbom, Master. Gilbert Sigvard Emanuel    male  0.33   
1240       3         1          Thomas, Master. Assad Alexander    male  0.42   
657        3         1                   Baclini, Miss. Eugenie  female  0.75   
658        3         1            Baclini, Miss. Helene Barbara  female  0.75   

      sibsp  parch     ticket     fare cabin embarked boat  body  \
763       1      2  C.A. 2315  20.5750   NaN        S   10   NaN   
747       0      2     347080  14.4000   NaN        S  NaN   NaN   
1240      0      1       2625   8.5167   NaN        C   16   NaN   
657       2      1       2666  19.2583   NaN        C    C   NaN   
658       2      1       2666  19.2583   NaN        C    C   NaN   

                       home.dest  modified_fare  
763   Devon, England Wichita, KS        2.05750  
747                  Stanton, IA        1.44000  
1240                         NaN        0.85167  
657           Syria New York, NY        1.92583  
658           Syria New York, NY        1.92583

In [189]:
# notice the difference

titanic.sort_values(by=['age', 'pclass'], ascending=[True, False]).head()

pclass  survived                                     name     sex   age  \
763        3         1  Dean, Miss. Elizabeth Gladys "Millvina"  female  0.17   
747        3         0  Danbom, Master. Gilbert Sigvard Emanuel    male  0.33   
1240       3         1          Thomas, Master. Assad Alexander    male  0.42   
427        2         1                Hamalainen, Master. Viljo    male  0.67   
657        3         1                   Baclini, Miss. Eugenie  female  0.75   

      sibsp  parch     ticket     fare cabin embarked boat  body  \
763       1      2  C.A. 2315  20.5750   NaN        S   10   NaN   
747       0      2     347080  14.4000   NaN        S  NaN   NaN   
1240      0      1       2625   8.5167   NaN        C   16   NaN   
427       1      1     250649  14.5000   NaN        S    4   NaN   
657       2      1       2666  19.2583   NaN        C    C   NaN   

                       home.dest  modified_fare  
763   Devon, England Wichita, KS        2.05750  
747                  Stanton, IA        1.44000  
1240                         NaN        0.85167  
427                  Detroit, MI        1.45000  
657           Syria New York, NY        1.92583

#### Replace Empty Values

Another way of dealing with empty cells is to insert a new value instead.

This way you do not have to delete entire rows just because of some empty cells.

The fillna() method allows us to replace empty cells with a value:

In [190]:
#Replace NULL values with the number 130:

df_path = r"C:\Users\12489\Documents\Data\lecture_270222.csv"
df = pd.read_csv(df_path)
df.info()

# df['Calories'].fillna(130, inplace = True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 169 entries, 0 to 168
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  169 non-null    int64  
 1   Pulse     169 non-null    int64  
 2   Maxpulse  169 non-null    int64  
 3   Calories  164 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 5.4 KB


Replace the missing value: constant, mean or median

In [191]:
# using mean
x = df['Calories'].mean()
df["Calories"].fillna(x, inplace = True)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 169 entries, 0 to 168
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  169 non-null    int64  
 1   Pulse     169 non-null    int64  
 2   Maxpulse  169 non-null    int64  
 3   Calories  169 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 5.4 KB


In [192]:
# using median
df = pd.read_csv(df_path)

x = df['Calories'].median()
df["Calories"].fillna(x, inplace = True)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 169 entries, 0 to 168
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  169 non-null    int64  
 1   Pulse     169 non-null    int64  
 2   Maxpulse  169 non-null    int64  
 3   Calories  169 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 5.4 KB


In [193]:
#using mode
df = pd.read_csv(df_path)

x = df["Calories"].mode()[0]
df["Calories"].fillna(x, inplace = True)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 169 entries, 0 to 168
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  169 non-null    int64  
 1   Pulse     169 non-null    int64  
 2   Maxpulse  169 non-null    int64  
 3   Calories  169 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 5.4 KB


# Missing value imputation using Simple Imputer
1. drop all records where missing value is present: stock market price, loss of records will me bare min
2. impute them or replace with some  metric like mean

In [194]:
flight_path = r"C:\Users\12489\Documents\Data\flights.csv"
df = pd.read_csv(flight_path)

df.tail()

mon  dom  dow carrier  flight  org  mile  depart  duration  delay
49995    9    8    3      OO    6529  ORD   109   13.08        55   40.0
49996    1   27    3      AA     194  SFO  2704   14.42       325   24.0
49997   10   21    5      UA     520  ORD   783    8.23       124  -13.0
49998    3   20    0      UA     835  LGA   733    8.00       150   25.0
49999    1   17    0      UA     766  ORD   622   21.00       103   19.0

In [195]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 10 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   mon       50000 non-null  int64  
 1   dom       50000 non-null  int64  
 2   dow       50000 non-null  int64  
 3   carrier   50000 non-null  object 
 4   flight    50000 non-null  int64  
 5   org       50000 non-null  object 
 6   mile      50000 non-null  int64  
 7   depart    50000 non-null  float64
 8   duration  50000 non-null  int64  
 9   delay     47022 non-null  float64
dtypes: float64(2), int64(6), object(2)
memory usage: 3.8+ MB


In [196]:
df.isna().any() #same as isnull

mon         False
dom         False
dow         False
carrier     False
flight      False
org         False
mile        False
depart      False
duration    False
delay        True
dtype: bool

In [197]:
#drop columns
df = df.drop(columns=['flight', 'carrier', 'org'])

In [198]:
df.head()

mon  dom  dow  mile  depart  duration  delay
0   11   20    6  2153    9.48       351    NaN
1    0   22    2   316   16.33        82   30.0
2    2   20    4   337    6.17        82   -8.0
3    9   13    1  1236   10.33       195   -5.0
4    4    2    5   258    8.92        65    NaN

In [199]:
# dont forget: "zero" might mean Null

df['duration'].isin([0]).any()

False

In [200]:
df['dow'].value_counts()

# but in THIS case, "zero" means Monday

5    7900
1    7441
2    7286
4    7265
0    7072
3    7057
6    5979
Name: dow, dtype: int64

In [201]:
import numpy as np

#df.dow.replace(0, np.nan, inplace=True)

#df.info()

### even though the professor wants me to drop,
### im not going to argue about "Monday"

In [202]:
#dropping all the rows where missing values are present

df_complete = df.dropna()
df_complete.shape

(47022, 7)

In [203]:
df.shape

# proof: hasnt changed

(50000, 7)

In [204]:
# imputation 
# (replacement of null values)
from sklearn.impute import SimpleImputer

# make a copy of my data
df_median = df.copy(deep=True)

#create a median imputer object
median_imputer = SimpleImputer(strategy='median')

#impute median values in the datframe df_median
df_median['delay'] = median_imputer.fit_transform(df_median['delay'].to_numpy().reshape(-1,1))

df_median.head()

mon  dom  dow  mile  depart  duration  delay
0   11   20    6  2153    9.48       351   16.0
1    0   22    2   316   16.33        82   30.0
2    2   20    4   337    6.17        82   -8.0
3    9   13    1  1236   10.33       195   -5.0
4    4    2    5   258    8.92        65   16.0

### imputer really SHINES when applying to ENTIRE df

In [205]:
# make a copy of my data
df_mean = df.copy(deep=True)

#create a median imputer object
mean_imputer = SimpleImputer(strategy='mean')

#impute mean values in the datframe df_mean
df_mean.iloc[:,:] = mean_imputer.fit_transform(df_mean)

df_mean.head()

mon   dom  dow    mile  depart  duration      delay
0  11.0  20.0  6.0  2153.0    9.48     351.0  28.663796
1   0.0  22.0  2.0   316.0   16.33      82.0  30.000000
2   2.0  20.0  4.0   337.0    6.17      82.0  -8.000000
3   9.0  13.0  1.0  1236.0   10.33     195.0  -5.000000
4   4.0   2.0  5.0   258.0    8.92      65.0  28.663796

In [206]:
#make a copy of my data
df_mode = df.copy(deep=True)

#create a median imputer object
mode_imputer = SimpleImputer(strategy='most_frequent')

#impute median values in the datframe df_median
df_mode.iloc[:,:] = mode_imputer.fit_transform(df_mode)

df_mode.head()

mon   dom  dow    mile  depart  duration  delay
0  11.0  20.0  6.0  2153.0    9.48     351.0   -7.0
1   0.0  22.0  2.0   316.0   16.33      82.0   30.0
2   2.0  20.0  4.0   337.0    6.17      82.0   -8.0
3   9.0  13.0  1.0  1236.0   10.33     195.0   -5.0
4   4.0   2.0  5.0   258.0    8.92      65.0   -7.0

In [207]:
#make a copy of my data
df_constant = df.copy(deep=True)

#create a constant imputer object
constant_imputer = SimpleImputer(strategy='constant', fill_value=0)

#impute constant values in the dataframe
df_constant.iloc[:,:] = constant_imputer.fit_transform(df_constant)

df_constant.head()

mon   dom  dow    mile  depart  duration  delay
0  11.0  20.0  6.0  2153.0    9.48     351.0    0.0
1   0.0  22.0  2.0   316.0   16.33      82.0   30.0
2   2.0  20.0  4.0   337.0    6.17      82.0   -8.0
3   9.0  13.0  1.0  1236.0   10.33     195.0   -5.0
4   4.0   2.0  5.0   258.0    8.92      65.0    0.0

In [208]:
# concatenation using pandas

na_path = r"C:\Users\12489\Documents\Data\north_america_2000_2010.csv"
sa_path = r"C:\Users\12489\Documents\Data\south_america_2000_2010.csv"

north_america = pd.read_csv(na_path, index_col=0)
south_america = pd.read_csv(sa_path, index_col=0)

In [209]:
north_america.tail()

2000    2001    2002    2003    2004  2005    2006    2007  2008  \
Country                                                                       
Canada   1779.0  1771.0  1754.0  1740.0  1760.0  1747  1745.0  1741.0  1735   
Mexico   2311.2  2285.2  2271.2  2276.5  2270.6  2281  2280.6  2261.4  2258   
USA      1836.0  1814.0  1810.0  1800.0  1802.0  1799  1800.0  1798.0  1792   

           2009    2010  
Country                  
Canada   1701.0  1703.0  
Mexico   2250.2  2242.4  
USA      1767.0  1778.0

In [210]:
south_america.tail()

2000  2001  2002  2003  2004  2005  2006  2007  2008  2009    2010
Country                                                                    
Chile    2263  2242  2250  2235  2232  2157  2165  2128  2095  2074  2069.6

In [211]:
americas = pd.concat([north_america, south_america], axis=0)

americas.tail()

# axis = 1 
# ^^ if, (for example), data on EXACT SAME russian oblasts
# from 1990 - 2001 & then 2002 - 2020 ^^

2000    2001    2002    2003    2004  2005    2006    2007  2008  \
Country                                                                       
Canada   1779.0  1771.0  1754.0  1740.0  1760.0  1747  1745.0  1741.0  1735   
Mexico   2311.2  2285.2  2271.2  2276.5  2270.6  2281  2280.6  2261.4  2258   
USA      1836.0  1814.0  1810.0  1800.0  1802.0  1799  1800.0  1798.0  1792   
Chile    2263.0  2242.0  2250.0  2235.0  2232.0  2157  2165.0  2128.0  2095   

           2009    2010  
Country                  
Canada   1701.0  1703.0  
Mexico   2250.2  2242.4  
USA      1767.0  1778.0  
Chile    2074.0  2069.6

### Adding Columns

In [212]:
# Define a dictionary containing Students data
data = {'Name': ['Jai', 'Princi', 'Gaurav', 'Anuj'],
        'Height': [5.1, 6.2, 5.1, 5.2],
        'Qualification': ['Msc', 'MA', 'Msc', 'Msc']}
 
# Convert the dictionary into DataFrame
df = pd.DataFrame(data)
 
# Declare a list (address) to be converted to column
address = ['Delhi', 'Bangalore', 'Chennai', 'Patna']
 
# Use 'Address' as the column name & add it to the df
df['Address'] = address

# Observe the result
df

Name  Height Qualification    Address
0     Jai     5.1           Msc      Delhi
1  Princi     6.2            MA  Bangalore
2  Gaurav     5.1           Msc    Chennai
3    Anuj     5.2           Msc      Patna

In [213]:
# Reset the dataframe
df = pd.DataFrame(data)
 
# Define a dictionary with key values of
# an existing column and their respective
# value pairs as the # values for our new column.

address_dict = {'Delhi':'Jai', 'Bangalore':'Princi', 'Patna':'Gaurav', 'Chennai':'Anuj'}
 
# add the dictionary to DataFrame
df['Address']=address_dict
 
# Observe the output
df

Name  Height Qualification    Address
0     Jai     5.1           Msc      Delhi
1  Princi     6.2            MA  Bangalore
2  Gaurav     5.1           Msc      Patna
3    Anuj     5.2           Msc    Chennai

In [214]:
# Reset the data
df = pd.DataFrame(data)
 
# Using DataFrame.insert() to add a column
df.insert(2, 'Age', [21,22,23,24], True)
"""index, name, values, allow duplicate?"""
 
# Observe the result
df

Name  Height  Age Qualification
0     Jai     5.1   21           Msc
1  Princi     6.2   22            MA
2  Gaurav     5.1   23           Msc
3    Anuj     5.2   24           Msc

### Reindexing

In [215]:
import numpy as np
 
index = ['Firefox', 'Chrome', 'Safari', 'IE10', 'Konqueror']
df = pd.DataFrame({'http_status': [200, 200, 404, 404, 301],
                  'response_time': [0.04, 0.02, 0.07, 0.08, 1.0]},
                  index=index)

new_index = ['NEW Safari', 'Iceweasel', 'Comodo Dragon', 'IE10',
             'Chrome']
df.reindex(new_index)

http_status  response_time
NEW Safari             NaN            NaN
Iceweasel              NaN            NaN
Comodo Dragon          NaN            NaN
IE10                 404.0           0.08
Chrome               200.0           0.02

In [216]:
# Reset the dataframe 
df1 = pd.DataFrame(np.random.rand(5,5),
           columns=column, index=index)
 
column2=['e','a','b','c','d']
  
# create the new index for columns
df1.reindex(column2, axis='columns')

e         a         b         c         d
Firefox    0.266251  0.837842  0.551620  0.618949  0.703681
Chrome     0.312667  0.746043  0.093183  0.621395  0.933999
Safari     0.296356  0.578344  0.292692  0.354329  0.186819
IE10       0.100954  0.906292  0.235899  0.914209  0.739439
Konqueror  0.000181  0.129516  0.515380  0.232284  0.792445

## Merge

In [217]:
temp_path = r"C:\Users\12489\Documents\Data\climate_temp.csv"
precip_path = r"C:\Users\12489\Documents\Data\climate_precip.csv"


climate_temp = pd.read_csv(temp_path)
climate_precip = pd.read_csv(precip_path)

In [218]:
climate_temp.head()

STATION            STATION_NAME ELEVATION  LATITUDE   LONGITUDE  \
0  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
1  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
2  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
3  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
4  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   

       DATE  DLY-CLDD-BASE45  DLY-CLDD-BASE50  DLY-CLDD-BASE55  \
0  20100101                6                2            -7777   
1  20100102                6                2                1   
2  20100103                6                2                1   
3  20100104                6                2                1   
4  20100105                6                2                1   

   DLY-CLDD-BASE57  ...  DLY-CLDD-NORMAL  DLY-CLDD-BASE70  DLY-CLDD-BASE72  \
0            -7777  ...                0                0                0   
1            -7777  ...                0                0                0   
2            -7777  ...                0                0                0   
3            -7777  ...                0                0                0   
4            -7777  ...                0                0                0   

   DLY-HTDD-BASE40  DLY-HTDD-BASE45  DLY-HTDD-BASE50  DLY-HTDD-BASE55  \
0            -7777                1                2                6   
1            -7777                1                2                6   
2            -7777                1                2                5   
3            -7777                1                2                5   
4            -7777            -7777                2                5   

   DLY-HTDD-BASE57  DLY-HTDD-BASE60  DLY-HTDD-NORMAL  
0                7               10               15  
1                7               10               15  
2                7               10               15  
3                7               10               15  
4                7               10               15  

[5 rows x 21 columns]

In [219]:
climate_precip.head()

STATION            STATION_NAME      DATE  DLY-PRCP-25PCTL  \
0  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100101            -6.66   
1  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100102            -6.66   
2  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100103            -6.66   
3  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100104            -6.66   
4  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100105            -6.66   

   DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
0             -666            -66.6            -6.66             -666   
1             -666            -66.6            -6.66             -666   
2             -666            -66.6            -6.66             -666   
3            -9999          -9999.0            -6.66            -9999   
4            -9999          -9999.0            -6.66            -9999   

   DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  ...  DLY-PRCP-PCTALL-GE100HI  \
0            -66.6            -6.66  ...                        3   
1            -66.6            -6.66  ...                        3   
2            -66.6            -6.66  ...                        3   
3          -9999.0            -6.66  ...                        3   
4          -9999.0            -6.66  ...                        3   

   DLY-SNWD-PCTALL-GE001WI  DLY-SNWD-PCTALL-GE010WI  DLY-SNWD-PCTALL-GE003WI  \
0                    -9999                        0                    -9999   
1                    -9999                        0                    -9999   
2                    -9999                        0                    -9999   
3                        0                        0                        0   
4                        0                        0                        0   

   DLY-SNWD-PCTALL-GE005WI  DLY-SNOW-PCTALL-GE001TI  DLY-SNOW-PCTALL-GE010TI  \
0                    -9999                    -9999                    -9999   
1                    -9999                    -9999                    -9999   
2                    -9999                    -9999                    -9999   
3                        0                        0                        0   
4                        0                        0                        0   

   DLY-SNOW-PCTALL-GE100TI  DLY-SNOW-PCTALL-GE030TI  DLY-SNOW-PCTALL-GE050TI  
0                        0                    -9999                    -9999  
1                        0                    -9999                    -9999  
2                        0                    -9999                    -9999  
3                        0                        0                        0  
4                        0                        0                        0  

[5 rows x 29 columns]

In [220]:
climate_temp.shape

(127020, 21)

In [221]:
climate_precip.shape

(151110, 29)

### Inner Join

Here you have created a new DataFrame called precip_one_station from the climate_precip DataFrame, selecting only rows in which the STATION field is "GHCND:USC00045721".

In [222]:
precip_one_station = climate_precip[climate_precip['STATION'] == "GHCND:USC00045721"]

precip_one_station.shape

(365, 29)

In [223]:
inner_merged = pd.merge(precip_one_station, climate_temp)

inner_merged.shape

(365, 47)

If you guessed 365 rows, then you were correct! This is because merge() defaults to an inner join, and an inner join will discard only those rows that do not match. Since all of your rows had a match, none were lost. You should also notice that there are many more columns now: 47 to be exact.

With merge(), you also have control over which column(s) to join on. Let’s say you want to merge both entire datasets, but only on Station and Date since the combination of the two will yield a unique value for each row. To do so, you can use the on parameter:

In [224]:
inner_merged_total = pd.merge(climate_temp, climate_precip, )

inner_merged_total.head()

STATION            STATION_NAME ELEVATION  LATITUDE   LONGITUDE  \
0  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
1  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
2  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
3  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
4  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   

       DATE  DLY-CLDD-BASE45  DLY-CLDD-BASE50  DLY-CLDD-BASE55  \
0  20100101                6                2            -7777   
1  20100102                6                2                1   
2  20100103                6                2                1   
3  20100104                6                2                1   
4  20100105                6                2                1   

   DLY-CLDD-BASE57  ...  DLY-PRCP-PCTALL-GE100HI  DLY-SNWD-PCTALL-GE001WI  \
0            -7777  ...                        3                    -9999   
1            -7777  ...                        3                    -9999   
2            -7777  ...                        3                    -9999   
3            -7777  ...                        3                        0   
4            -7777  ...                        3                        0   

   DLY-SNWD-PCTALL-GE010WI  DLY-SNWD-PCTALL-GE003WI  DLY-SNWD-PCTALL-GE005WI  \
0                        0                    -9999                    -9999   
1                        0                    -9999                    -9999   
2                        0                    -9999                    -9999   
3                        0                        0                        0   
4                        0                        0                        0   

   DLY-SNOW-PCTALL-GE001TI  DLY-SNOW-PCTALL-GE010TI  DLY-SNOW-PCTALL-GE100TI  \
0                    -9999                    -9999                        0   
1                    -9999                    -9999                        0   
2                    -9999                    -9999                        0   
3                        0                        0                        0   
4                        0                        0                        0   

   DLY-SNOW-PCTALL-GE030TI  DLY-SNOW-PCTALL-GE050TI  
0                    -9999                    -9999  
1                    -9999                    -9999  
2                    -9999                    -9999  
3                        0                        0  
4                        0                        0  

[5 rows x 47 columns]

In [225]:
inner_merged_total.shape

(123005, 47)

You can specify a single key column with a string or multiple key columns with a list. This results in a DataFrame with 123,005 rows and 48 columns.

Why 48 columns instead of 47? Because you specified the key columns to join on, Pandas doesn’t try to merge all mergeable columns. This can result in “duplicate” column names, which may or may not have different values.

“Duplicate” is in quotes because the column names will not be an exact match. By default they are appended with _x and _y

With an outer join, you can expect to have the same number of rows as the larger DataFrame. That’s because no rows are lost in an outer join, even when they don’t have a match in the other DataFrame.

Left Join In this example, you’ll specify a left join—also known as a left outer join—with the how parameter. Using a left outer join will leave your new merged DataFrame with all rows from the left DataFrame, while discarding rows from the right DataFrame that don’t have a match in the key column of the left DataFrame.

In [226]:
left_merged = pd.merge(climate_temp, precip_one_station,
                        how="left", on=["STATION", "DATE"])

left_merged.head()

STATION          STATION_NAME_x ELEVATION  LATITUDE   LONGITUDE  \
0  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
1  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
2  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
3  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
4  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   

       DATE  DLY-CLDD-BASE45  DLY-CLDD-BASE50  DLY-CLDD-BASE55  \
0  20100101                6                2            -7777   
1  20100102                6                2                1   
2  20100103                6                2                1   
3  20100104                6                2                1   
4  20100105                6                2                1   

   DLY-CLDD-BASE57  ...  DLY-PRCP-PCTALL-GE100HI  DLY-SNWD-PCTALL-GE001WI  \
0            -7777  ...                      NaN                      NaN   
1            -7777  ...                      NaN                      NaN   
2            -7777  ...                      NaN                      NaN   
3            -7777  ...                      NaN                      NaN   
4            -7777  ...                      NaN                      NaN   

   DLY-SNWD-PCTALL-GE010WI  DLY-SNWD-PCTALL-GE003WI  DLY-SNWD-PCTALL-GE005WI  \
0                      NaN                      NaN                      NaN   
1                      NaN                      NaN                      NaN   
2                      NaN                      NaN                      NaN   
3                      NaN                      NaN                      NaN   
4                      NaN                      NaN                      NaN   

   DLY-SNOW-PCTALL-GE001TI  DLY-SNOW-PCTALL-GE010TI  DLY-SNOW-PCTALL-GE100TI  \
0                      NaN                      NaN                      NaN   
1                      NaN                      NaN                      NaN   
2                      NaN                      NaN                      NaN   
3                      NaN                      NaN                      NaN   
4                      NaN                      NaN                      NaN   

   DLY-SNOW-PCTALL-GE030TI  DLY-SNOW-PCTALL-GE050TI  
0                      NaN                      NaN  
1                      NaN                      NaN  
2                      NaN                      NaN  
3                      NaN                      NaN  
4                      NaN                      NaN  

[5 rows x 48 columns]

In [227]:
left_merged.shape

(127020, 48)

Right Join The right join (or right outer join) is the mirror-image version of the left join. With this join, all rows from the right DataFrame will be retained, while rows in the left DataFrame without a match in the key column of the right DataFrame will be discarded.

To demonstrate how right and left joins are mirror images of each other, in the example below you’ll recreate the left_merged DataFrame from above, only this time using a right join:

In [228]:
right_merged = pd.merge(precip_one_station, climate_temp, how="right", on=["STATION", "DATE"])
right_merged.head()

STATION STATION_NAME_x      DATE  DLY-PRCP-25PCTL  \
0  GHCND:USC00049099            NaN  20100101              NaN   
1  GHCND:USC00049099            NaN  20100102              NaN   
2  GHCND:USC00049099            NaN  20100103              NaN   
3  GHCND:USC00049099            NaN  20100104              NaN   
4  GHCND:USC00049099            NaN  20100105              NaN   

   DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
0              NaN              NaN              NaN              NaN   
1              NaN              NaN              NaN              NaN   
2              NaN              NaN              NaN              NaN   
3              NaN              NaN              NaN              NaN   
4              NaN              NaN              NaN              NaN   

   DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  ...  DLY-CLDD-NORMAL  DLY-CLDD-BASE70  \
0              NaN              NaN  ...                0                0   
1              NaN              NaN  ...                0                0   
2              NaN              NaN  ...                0                0   
3              NaN              NaN  ...                0                0   
4              NaN              NaN  ...                0                0   

   DLY-CLDD-BASE72  DLY-HTDD-BASE40  DLY-HTDD-BASE45  DLY-HTDD-BASE50  \
0                0            -7777                1                2   
1                0            -7777                1                2   
2                0            -7777                1                2   
3                0            -7777                1                2   
4                0            -7777            -7777                2   

   DLY-HTDD-BASE55  DLY-HTDD-BASE57  DLY-HTDD-BASE60  DLY-HTDD-NORMAL  
0                6                7               10               15  
1                6                7               10               15  
2                5                7               10               15  
3                5                7               10               15  
4                5                7               10               15  

[5 rows x 48 columns]

In [229]:
right_merged.shape

(127020, 48)